File: TerminalVoiceOverlay-macOS/TerminalVoiceOverlay/RecordingArchive.swift
import Foundation

/// Bewahrt die zuletzt aufgenommenen Diktate auf, statt sie nach dem Turn zu loeschen.
///
/// Vorfall 29.08.2026 (Windows-Schwester-App): Ein 15,4-Minuten-Diktat ergab eine 29,5-MB-WAV,
/// die Groq mit HTTP 413 abwies; die Datei wurde danach geloescht — der gesprochene Text war
/// endgueltig weg und musste komplett neu gesprochen werden. Seitdem gilt: **eine Aufnahme wird
/// nie mehr direkt geloescht**. Jede Aufnahme wandert hierher, die letzten `keepCount` bleiben
/// erhalten und koennen ueber Rechtsklick auf das Mikrofon erneut transkribiert werden.
///
/// Bewusst klein gehalten: reines Dateisystem, kein Zustand im Speicher, damit das Archiv auch
/// nach einem Absturz der App vollstaendig ist. 1:1-Pendant zu `RecordingArchive.cs` unter Windows.
enum RecordingArchive {

    /// So viele Aufnahmen bleiben erhalten (die letzte und die vorletzte).
    static let keepCount = 2

    private static let filePrefix = "aufnahme_"

    /// ~/Library/Application Support/TerminalVoiceOverlay/aufnahmen
    static var directory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent("Library/Application Support")
        let dir = base
            .appendingPathComponent("TerminalVoiceOverlay", isDirectory: true)
            .appendingPathComponent("aufnahmen", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    /// Verschiebt eine frisch aufgenommene WAV ins Archiv und raeumt alles ausser den letzten
    /// `keepCount` Aufnahmen ab. Gibt den neuen Pfad zurueck (nil, wenn nichts zu archivieren war).
    /// Wirft nie — das Archiv darf einen Turn niemals zum Scheitern bringen.
    @discardableResult
    static func archive(_ fileURL: URL?, success: Bool) -> URL? {
        guard let fileURL = fileURL else { return nil }
        let fm = FileManager.default
        guard fm.fileExists(atPath: fileURL.path) else { return nil }

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd_HH-mm-ss"
        let stamp = formatter.string(from: Date())
        let status = success ? "ok" : "fehler"

        var target = directory.appendingPathComponent("\(filePrefix)\(stamp)_\(status).wav")
        // Zwei Aufnahmen in derselben Sekunde: Zaehler anhaengen statt ueberschreiben.
        var counter = 2
        while fm.fileExists(atPath: target.path) && counter < 100 {
            target = directory.appendingPathComponent("\(filePrefix)\(stamp)_\(status)_\(counter).wav")
            counter += 1
        }

        do {
            try fm.moveItem(at: fileURL, to: target)
            let attrs = try? fm.attributesOfItem(atPath: target.path)
            let bytes = (attrs?[.size] as? NSNumber)?.intValue ?? 0
            DiagLog.write("Recording", "archived",
                          [("path", target.path), ("bytes", bytes), ("success", success)])
            prune()
            return target
        } catch {
            DiagLog.warn("Recording", "archive_failed", [("err", error.localizedDescription)])
            return nil
        }
    }

    /// Die aufbewahrten Aufnahmen, neueste zuerst (hoechstens `keepCount`).
    static func recent() -> [URL] {
        let fm = FileManager.default
        guard let files = try? fm.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else { return [] }

        return files
            .filter { $0.lastPathComponent.hasPrefix(filePrefix) && $0.pathExtension == "wav" }
            .sorted { modified($0) > modified($1) }
            .prefix(keepCount)
            .map { $0 }
    }

    /// Die n-letzte Aufnahme (0 = die letzte, 1 = die vorletzte) oder nil, wenn es sie nicht gibt.
    static func at(_ index: Int) -> URL? {
        let all = recent()
        return index >= 0 && index < all.count ? all[index] : nil
    }

    /// Loescht alles ausser den letzten `keepCount` Aufnahmen.
    private static func prune() {
        let fm = FileManager.default
        guard let files = try? fm.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else { return }

        let stale = files
            .filter { $0.lastPathComponent.hasPrefix(filePrefix) && $0.pathExtension == "wav" }
            .sorted { modified($0) > modified($1) }
            .dropFirst(keepCount)

        for old in stale {
            do {
                try fm.removeItem(at: old)
                DiagLog.write("Recording", "archive_pruned", [("path", old.path)])
            } catch {
                // Gesperrt -> die naechste Aufnahme raeumt sie ab.
            }
        }
    }

    private static func modified(_ url: URL) -> Date {
        (try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
    }
}

File: TerminalVoiceOverlay-Windows/Services/RecordingArchive.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace TerminalVoiceOverlay.Services
{
    /// <summary>
    /// Bewahrt die zuletzt aufgenommenen Diktate auf, statt sie nach dem Turn zu loeschen.
    ///
    /// Vorfall 29.08.2026: Ein 15,4-Minuten-Diktat ergab eine 29,5-MB-WAV, die Groq mit HTTP 413
    /// abwies; die Datei wurde danach im <c>finally</c> geloescht — der gesprochene Text war
    /// endgueltig weg und musste komplett neu gesprochen werden. Seitdem gilt: <b>eine Aufnahme
    /// wird nie mehr direkt geloescht</b>. Jede Aufnahme wandert hierher, die letzten
    /// <see cref="KeepCount"/> bleiben erhalten und koennen ueber Rechtsklick auf das Mikrofon
    /// erneut transkribiert werden. Erst die uebernaechste Aufnahme raeumt die aelteste ab.
    ///
    /// Bewusst klein gehalten: reines Dateisystem, kein Zustand im Speicher, damit das Archiv auch
    /// nach einem Absturz der App vollstaendig ist.
    /// </summary>
    public static class RecordingArchive
    {
        /// <summary>So viele Aufnahmen bleiben erhalten (die letzte und die vorletzte).</summary>
        public const int KeepCount = 2;

        private const string FilePrefix = "aufnahme_";

        /// <summary>%LOCALAPPDATA%\TerminalVoiceOverlay\aufnahmen — Ablage der letzten Diktate.</summary>
        public static string Directory
        {
            get
            {
                string dir = Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                    "TerminalVoiceOverlay", "aufnahmen");
                System.IO.Directory.CreateDirectory(dir);
                return dir;
            }
        }

        /// <summary>
        /// Verschiebt eine frisch aufgenommene WAV ins Archiv und raeumt alles ausser den letzten
        /// <see cref="KeepCount"/> Aufnahmen ab. Gibt den neuen Pfad zurueck (null, wenn nichts zu
        /// archivieren war). Wirft nie — das Archiv darf einen Turn niemals zum Scheitern bringen.
        /// </summary>
        /// <param name="wavFile">Pfad der Aufnahme im Temp-Ordner.</param>
        /// <param name="success">Ob der Turn erfolgreich war (steht nur im Dateinamen).</param>
        public static string? Archive(string? wavFile, bool success)
        {
            try
            {
                if (string.IsNullOrEmpty(wavFile) || !File.Exists(wavFile)) return null;

                string dir = Directory;
                string stamp = DateTime.Now.ToString("yyyy-MM-dd_HH-mm-ss");
                string status = success ? "ok" : "fehler";
                string target = Path.Combine(dir, $"{FilePrefix}{stamp}_{status}.wav");

                // Zwei Aufnahmen in derselben Sekunde: Zaehler anhaengen statt ueberschreiben.
                for (int i = 2; File.Exists(target) && i < 100; i++)
                    target = Path.Combine(dir, $"{FilePrefix}{stamp}_{status}_{i}.wav");

                File.Move(wavFile, target, overwrite: false);
                long bytes = new FileInfo(target).Length;
                DiagLog.Write("Recording", "archived",
                    ("path", target), ("bytes", bytes), ("success", success));

                Prune(dir);
                return target;
            }
            catch (Exception ex)
            {
                DiagLog.Warn("Recording", "archive_failed",
                    ("err", ex.Message), ("type", ex.GetType().Name));
                return null;
            }
        }

        /// <summary>
        /// Die aufbewahrten Aufnahmen, neueste zuerst (hoechstens <see cref="KeepCount"/>).
        /// </summary>
        public static IReadOnlyList<string> Recent()
        {
            try
            {
                return System.IO.Directory
                    .EnumerateFiles(Directory, FilePrefix + "*.wav")
                    .OrderByDescending(File.GetLastWriteTimeUtc)
                    .Take(KeepCount)
                    .ToList();
            }
            catch (Exception ex)
            {
                DiagLog.Warn("Recording", "archive_list_failed",
                    ("err", ex.Message), ("type", ex.GetType().Name));
                return Array.Empty<string>();
            }
        }

        /// <summary>
        /// Die n-letzte Aufnahme (0 = die letzte, 1 = die vorletzte) oder null, wenn es sie nicht gibt.
        /// </summary>
        public static string? At(int index)
        {
            var all = Recent();
            return index >= 0 && index < all.Count ? all[index] : null;
        }

        /// <summary>Loescht alles ausser den letzten <see cref="KeepCount"/> Aufnahmen.</summary>
        private static void Prune(string dir)
        {
            try
            {
                var stale = System.IO.Directory
                    .EnumerateFiles(dir, FilePrefix + "*.wav")
                    .OrderByDescending(File.GetLastWriteTimeUtc)
                    .Skip(KeepCount)
                    .ToList();

                foreach (var old in stale)
                {
                    try
                    {
                        File.Delete(old);
                        DiagLog.Write("Recording", "archive_pruned", ("path", old));
                    }
                    catch { /* gesperrt -> die naechste Aufnahme raeumt sie ab */ }
                }
            }
            catch (Exception ex)
            {
                DiagLog.Warn("Recording", "archive_prune_failed",
                    ("err", ex.Message), ("type", ex.GetType().Name));
            }
        }
    }
}
